### PR TITLE
added metadata and tags functionality, and some tests

### PR DESCRIPTION
**Pull request by joe-sfx**
added python wrappers for most of the functionality supported by the SignalFx API
(https://developers.signalfx.com/): get_time_series_window;
searching and getting metadata (metrics, dimensions, metric time series);
searching, updating, getting, deleting tags

added unit tests for some of the new methods, and a "live" test which
establishes an API session and executes various sequences of the new
methods

### CODE REVIEW

**Comment by joe-sfx** [comment]
@mpetazzoni @ramjothikumar assigning to Mike, your comments would also be welcome

File: tests/live_tests.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
+"""
+"Live" tests for SignalFx Python client library.
+
+to run this script, use "python live_tests.py --token YOUR_TOKEN"
```
Comment (udaysagar2177) [inline tests/live_tests.py]: Do you want to include this circle.yml?

File: tests/live_tests.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+# import the correct signalfx
+this_path = os.path.dirname(os.path.realpath(__file__))
+desired_path = this_path.replace('tests', '')
+if desired_path not in sys.path: sys.path.insert(0, desired_path)
+import signalfx
+exec(CLIENT_NAME + '=signalfx.SignalFx(TOKEN)')
+
+def send_data_example(client_name, num_pts=10):
+    sess = eval(client_name)
+    init_stamp = 1453960607000
```
Comment (udaysagar2177) [inline tests/live_tests.py]: Consider using time.time()*1000?

File: tests/live_tests.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    send_data_example(CLIENT_NAME)
+    time.sleep(20)
+    apply_sequence(CLIENT_NAME, [gT, sT, dT, uT, gT, sT, dT, gT, sT, uT,
+                                 gT, sT, dT], (FAKE_TAG,), timeout=15)
+    test_func(CLIENT_NAME, sD, func_args=[FAKE_KEY], timeout=15)
+    apply_sequence(CLIENT_NAME, [uD, gD], (FAKE_KEY, FAKE_VALUE), timeout=15)
+    test_func(CLIENT_NAME, sD, func_args=[FAKE_VALUE], timeout=15)
+    test_func(CLIENT_NAME, 'get_time_series_window',
+              func_args=['sf_source:analytics-gauss AND sf_metric:"jvm.heap.used"',
+                         450000000000, 1455320231110], resolution=3600000, timeout=20)
```
Comment (udaysagar2177) [inline tests/live_tests.py]: In future, searching for this time range can be heavy. how about time.time()*1000 - (few seconds)?

File: signalfx/client.py [hunk 4]
```diff
@@ -232,27 +264,394 @@ def _send(self):
...
+        resp = self._get('{0}/{1}'.format(self._api_endpoint,
+                                          self._ORGANIZATION_ENDPOINT_SUFFIX),
+                         session=self._api_session, **kwargs)
+        resp.raise_for_status()
+        return resp.json()
+
+    def get_organization_member(self):
+        raise NotImplementedError
+
+    def send_organization_member(self):
```
Comment (udaysagar2177) [inline signalfx/client.py]: As these are not implemented yet, might as well remove them?

File: signalfx/client.py [hunk 5]
```diff
@@ -232,27 +264,394 @@ def _send(self):
...
+            resolution (optional[int]): resolution in milliseconds for data
+                returned, default=1000
+            timeout (optional[int]): how long to wait for response (in seconds)
+
+        Returns:
+            dictionary containing the data in the windows of the relevant time
+                series
+
+        """
+        # TO DO: refactor _search_metrics_and_metadata so this is special case
```
Comment (udaysagar2177) [inline signalfx/client.py]: Did you finish this todo?

File: signalfx/client.py [hunk 6]
```diff
@@ -232,27 +264,394 @@ def _send(self):
...
+
+        Args:
+            object_name (string): name of the object (e.g. 'jvm.cpu.load')
+            object_endpoint (string): API endpoint suffix (e.g. 'v2/tag')
+
+        Returns:
+            dictionary of response
+        """
+        # add this?
+        # logging.debug('Performing an elasticsearch for %(qry)s at %(pt)s',
```
Comment (udaysagar2177) [inline signalfx/client.py]: Check whole pull request and please remove this kind of comments

File: signalfx/client.py [hunk 7]
```diff
@@ -232,27 +264,394 @@ def _send(self):
...
+        response = _session.post(url, data=data, timeout=_timeout)
+        logging.debug('Sending to SignalFx %s (%d %s)',
+                      'succeeded' if response.ok else 'failed',
+                      response.status_code, response.text)
+
+    def _get(self, url, session=None, timeout=None):
+        _timeout = timeout or self._timeout
+        _session = session or self._api_session
+        logging.debug('url being gotten: %s', pprint.pformat(url))
+        response = _session.get(url, timeout=_timeout)
```
Comment (udaysagar2177) [inline signalfx/client.py]: May be surround with try and except with log messages?

File: tests/live_tests.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+def main():
+    send_data_example(CLIENT_NAME)
+    time.sleep(20)
+    apply_sequence(CLIENT_NAME, [gT, sT, dT, uT, gT, sT, dT, gT, sT, uT,
+                                 gT, sT, dT], (FAKE_TAG,), timeout=15)
+    test_func(CLIENT_NAME, sD, func_args=[FAKE_KEY], timeout=15)
+    apply_sequence(CLIENT_NAME, [uD, gD], (FAKE_KEY, FAKE_VALUE), timeout=15)
+    test_func(CLIENT_NAME, sD, func_args=[FAKE_VALUE], timeout=15)
+    test_func(CLIENT_NAME, 'get_time_series_window',
+              func_args=['sf_source:analytics-gauss AND sf_metric:"jvm.heap.used"',
```
Comment (udaysagar2177) [inline tests/live_tests.py]: User may not be sending this metric to search

File: signalfx/client.py (hunk 7 above)
Comment (joe-sfx) [inline signalfx/client.py]: refactored to be handled by the call to _post in _send

File: signalfx/client.py (hunk 5 above)
Comment (joe-sfx) [inline signalfx/client.py]: No.

File: signalfx/client.py (hunk 4 above)
Comment (joe-sfx) [inline signalfx/client.py]: Placeholders for the next round of work.   Will stash.

File: tests/live_tests.py (hunk 1 above)
Comment (joe-sfx) [inline tests/live_tests.py]: Yes.

File: tests/live_tests.py (hunk 2 above)
Comment (joe-sfx) [inline tests/live_tests.py]: Deleting since same test is done by test_data_reporting.py

File: signalfx/client.py [hunk 9]
```diff
@@ -241,18 +272,366 @@ def _send(self):
...
         try:
-            response = _session.post(url, data=data, timeout=self._timeout)
-            logging.debug('Sending to SignalFx %s (%d %s)',
-                          'succeeded' if response.ok else 'failed',
-                          response.status_code, response.text)
+            response = _session.post(url, data=data, timeout=_timeout)
         except Exception:
-            logging.exception('Posting to SignalFx failed.')
+            logging.exception('Posting data to SignalFx failed.')
+        logging.debug('Sending to SignalFx %s (%d %s)',
```
Comment (udaysagar2177) [inline signalfx/client.py]: this statement should be inside of try block. Incase of an exception, response is never assigned

File: signalfx/client.py [hunk 10]
```diff
@@ -241,18 +272,366 @@ def _send(self):
...
+        _session = session or self._api_session
+        logging.debug('url being gotten: %s', pprint.pformat(url))
+        try:
+            response = _session.get(url, timeout=_timeout)
+        except Exception:
+            logging.exception('Getting from SignalFx failed.')
+        logging.debug('Getting from SignalFx %s (%d %s)',
+                      'succeeded' if response.ok else 'failed',
+                      response.status_code, response.text)
+        return response
```
Comment (udaysagar2177) [inline signalfx/client.py]: response never assigned in case of an exception

File: signalfx/client.py [hunk 11]
```diff
@@ -241,18 +272,366 @@ def _send(self):
...
+
+    def _get(self, url, session=None, timeout=None):
+        _timeout = timeout or self._timeout
+        _session = session or self._api_session
+        logging.debug('url being gotten: %s', pprint.pformat(url))
+        try:
+            response = _session.get(url, timeout=_timeout)
+        except Exception:
+            logging.exception('Getting from SignalFx failed.')
+        logging.debug('Getting from SignalFx %s (%d %s)',
```
Comment (udaysagar2177) [inline signalfx/client.py]: ditto

File: signalfx/client.py [hunk 12]
```diff
@@ -241,18 +272,366 @@ def _send(self):
...
+            query)
+        url_to_get += '&startMs=' + str(start_ms)
+        url_to_get += '&endMs=' + str(end_ms)
+        # for resolution, use API default (by leaving out of url)
+        if resolution is not None:
+            url_to_get += '&resolution=' + str(resolution)
+        _timeout = timeout or self._timeout
+        resp = self._get(url_to_get, session=self._api_session,
+                         timeout=_timeout)
+        resp.raise_for_status()
```
Comment (udaysagar2177) [inline signalfx/client.py]: take care of this incase of an exception. resp may not have these methods raise_for_status() and json()

File: signalfx/client.py (hunk 9 above)
Comment (joe-sfx) [inline signalfx/client.py]: restoring old version which does not handle exceptions (i.e. leaves that to the user)

File: signalfx/client.py [hunk 13]
```diff
@@ -61,23 +62,48 @@ def send_event(self, event_type, category=None, dimensions=None,
...
+        return data
+
 
 class SignalFxClient(BaseSignalFx):
     """SignalFx API client.
 
     This class presents a programmatic interface to SignalFx's metadata and
-    ingest APIs. At the time being, only ingest is supported; more will come
-    later.
+    ingest APIs. At the time being, ingest and retrieval of data
```
Comment (mdubbyap) [inline signalfx/client.py]: NIT: Weird sentence structure "At the time being,..." At the moment?  For the time being?

File: signalfx/client.py [hunk 14]
```diff
@@ -61,23 +62,48 @@ def send_event(self, event_type, category=None, dimensions=None,
...
     _HEADER_USER_AGENT_KEY = 'User-Agent'
     _INGEST_ENDPOINT_DATAPOINT_SUFFIX = 'v2/datapoint'
     _INGEST_ENDPOINT_EVENT_SUFFIX = 'v2/event'
     _THREAD_NAME = 'SignalFxDatapointSendThread'
+    _METRIC_ENDPOINT_SUFFIX = 'v2/metric'
+    _DIMENSION_ENDPOINT_SUFFIX = 'v2/dimension'
+    _MTS_ENDPOINT_SUFFIX = 'v2/metrictimeseries'
+    _TAG_ENDPOINT_SUFFIX = 'v2/tag'
+    _ORGANIZATION_ENDPOINT_SUFFIX = 'v2/organization'
+    _TIME_SERIES_WINDOW_ENDPOINT_SUFFIX = 'v1/timeserieswindow'
```
Comment (mdubbyap) [inline signalfx/client.py]: are we sure we want to expose v1 apis?  @tedoc2000  i thought this wasn't something we wanted to do?

**Pull request merged**